zebra/zserv.c

```diff
@@ -483,27 +499,41 @@ static void zserv_client_event(struct zserv *client,
  * as the task argument.
  *
  * Each message is popped off the client's input queue and the action associated
- * with the message is executed. This proceeds until there are no more messages,
- * an error occurs, or the processing limit is reached.
+ * with the message is executed. This proceeds until an error occurs, or the
+ * processing limit is reached.
  *
  * The client's I/O thread can push at most zrouter.packets_to_process messages
  * onto the input buffer before notifying us there are packets to read. As long
  * as we always process zrouter.packets_to_process messages here, then we can
  * rely on the read thread to handle queuing this task enough times to process
  * everything on the input queue.
+ *
+ * In one iteration of wake up, the main thread can process at a maximum of
+ * configured ptp/4 items from the ibuf_fifo. If ibuf_fifo still has items in
+ * it, the main thread reschedules itself. This way we ensure that the main
+ * thread is always processing some packets.
+ *
+ * If the client ibuf has maximum items(packets_to_process) then main thread
+ * after processing ptp/4 items in the ibuf_fifo, it schedules a call to the
+ * client pthread to pump in more items on the client's input buffer fifo.
+ * This way we ensure that the client IO thread is always reading from socket
+ * buffer and adding items in the ibuf_fifo.
  */
 static void zserv_process_messages(struct event *thread)
 {
 	struct zserv *client = EVENT_ARG(thread);
 	struct stream *msg;
 	struct stream_fifo *cache = stream_fifo_new();
 	uint32_t p2p = zrouter.packets_to_process;
+	uint32_t p2p_limit = p2p / PTP_ZSERV_PROCESS_LIMIT;
```

Comment:
hmm, I wasn't really suggesting to change the work being done on the 'process' side - I just wanted to avoid a couple of situations that might come up. I think it's fine, in this initial version, to continue to do the same amount of work here.

Reply:
Uploaded new review based on our discussion